examples/step-86/step-86.cc

```diff
@@ -0,0 +1,896 @@
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/grid_out.h>
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/dofs/dof_tools.h>
+#include <deal.II/fe/fe_q.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/numerics/data_out.h>
+#include <deal.II/numerics/vector_tools.h>
+#include <deal.II/numerics/error_estimator.h>
+#include <deal.II/numerics/solution_transfer.h>
+#include <deal.II/numerics/matrix_tools.h>
+#include <deal.II/lac/petsc_ts.h>
+
+#include <fstream>
+#include <iostream>
+
+
+namespace Step86
+{
+  using namespace dealii;
+
+  // We derive from ParameterAcceptor to use the ParameterAcceptor facilities in
+  // order to parse and read options from a parameter file (see also step-70).
+  template <int dim>
+  class HeatEquation : public ParameterAcceptor
+  {
+  public:
+    HeatEquation(const MPI_Comm mpi_communicator = MPI_COMM_WORLD);
+    void run();
+
+  private:
+    // This is the usual function we need to setup dofs, constraints, and linear
+    // algebra objects.
+    void setup_system();
+
+    // Here is where we start to deviate from the "manual" way of solving time
+    // dependent problems. High level packages for the solution of ODEs and IDAs
+    // usually expect the user to provide a function that computes the residual
+    // w.r.t. the solution and the time derivative of the solution.
+    //
+    // This allows those packages to abstract away the details of how the time
+    // derivative is defined (e.g., backward Euler, Crank-Nicolson, etc.) and
+    // allow the user to provide a uniform interface, irrespective of the time
+    // stepper used. PETSc and SUNDIALS are two examples of such packages, and
+    // they require the user to provide C style call-backs to compute residuals,
+    // Jacobians, etc. In deal.II, we wrap these C stype libraries with our own
+    // wrappers, that use a style which is closer to c++, and that allows us to
+    // simply define the callbacks via lambda functions.
+    //
+    // To make it clear what we are doing here, we start by defining functions
+    // with the same name of the interface that we will use later on. These are
+    // the function that will be called by the PETSc time stepper, and consist
+    // in the residual function:
+    void implicit_function(const double                      time,
+                           const PETScWrappers::MPI::Vector &solution,
+                           const PETScWrappers::MPI::Vector &solution_dot,
+                           PETScWrappers::MPI::Vector &      residual) const;
+
+    // the Jacobian function
+    void
+    assemble_implicit_jacobian(const double                      time,
+                               const PETScWrappers::MPI::Vector &solution,
+                               const PETScWrappers::MPI::Vector &solution_dot,
+                               const double                      shift);
+
+    // the function that actually solves the linear system
```

Comment:
I would move this later on and mention the fact that this is not needed when we use `set_matrix` or `set_matrices`